pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,365 @@
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    "x": {(0,): [-0.5, 0.0, 0.5], (1,): [-0.5, 0.0, 0.5], (2,): [-0.5, 0.0, 0.5]}
+
+    Selecting arguments by name instead of index is possible via the
+    ``encoding_args`` argument:
+
+    >>> res = advanced_spectrum(circuit, encoding_args={"y"})(x, y, z, w)
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    "y": {(0,): [-2.3, 0.0, 2.3], (1,): [-2.3, 0.0, 2.3], (2,): [-2.3, 0.0, 2.3]}
+
+    Note that for array-valued arguments the spectrum for each element of the array
+    is computed. A more fine-grained control is available by passing index tuples
+    for the respective argument name in ``encoding_args``:
+
+    >>> encoding_args = {"y": [(0,),(2,)]}
+    >>> res = advanced_spectrum(circuit, encoding_args=encoding_args)(x, y, z, w)
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    "y": {(0,): [-2.3, 0.0, 2.3], (2,): [-2.3, 0.0, 2.3]}
+
+    .. warning::
+        The ``advanced_spectrum`` function does not check if the result of the
+        circuit is an expectation value. It checks whether the classical preprocessing between
+        QNode and gate arguments is linear by computing the Jacobian of the processing
+        at multiple points. This makes it unlikely -- *but not impossible* -- that
+        non-linear functions go undetected.
+        The number of additional points at which the Jacobian is computed can be controlled
+        via ``num_pos``, and the test is deactivated if ``num_pos=0`` (discouraged).
+        Furthermore, the QNode arguments *not* marked in ``argnum`` will not be
+        considered in this test and if they resemble encoded inputs, the entire
+        spectrum might be incorrect or the circuit might not even admit one.
+
+    The ``advanced_spectrum`` function works in all interfaces:
+
+    .. code-block:: python
+
+        import tensorflow as tf
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev, interface='tf')
+        def circuit(x):
+            qml.RX(0.4*x[0], wires=0)
+            qml.PhaseShift(x[1]*np.pi, wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        x = tf.constant([1., 2.])
+        res = advanced_spectrum(circuit)(x)
+
+    >>> print(res)
+    {"x": {(0,): [-0.4, 0.0, 0.4], (1,): [-3.14159, 0.0, 3.14159]}}
+
+    """
+    validation_kwargs = validation_kwargs or {}
+    encoding_args, argnum = _process_ids(encoding_args, argnum, qnode)
+    atol = 10 ** (-decimals) if decimals is not None else 1e-10
+    # A map between Jacobians (contiguous) and arg names (may be discontiguous)
+    arg_name_map = dict(enumerate(encoding_args))
+
+    @wraps(qnode)
+    def wrapper(*args, **kwargs):
+        # Compute classical Jacobian and assert preprocessing is linear
+        jac_fn = qml.transforms.classical_jacobian(qnode, argnum=argnum, expand_fn=expand_multi_par)
```

Comment:
since this line is independent of `args` and `kwargs`, you could move it out the wrapper function, to reduce some overhead

Reply:
Perfect! Thanks :)